replace deprecated QueryActorPublisher and AllPersistenceIdsPublisher

fancywriter opened the pull request.

fancywriter commented:
@patriknw @chbatey Hi guys, `Source.actorPublisher` is deprecated and `AllPersistenceIdsStage` looks much easier to read, at least for me. What do you think?

patriknw commented:
Thanks for PR, much appriciated. We will review. Have been busy and next week vacation, so have patience.

chbatey commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+                  query()
+
+                case Some(rs) =>
+                  if (rs.getExecutionInfo.getPagingState != null) {
```
I think this should check if the result set is exhausted, if so then kick off another query. If not and getAvailableWithoutFetching it can pull more off, otherwise it should kick off the fetchMoreResults

raboof replied:
This is now achieved by holding on to the `resultSet` and calling `fetchMoreResults` when demand is signalled while the buffer is empty.

fancywriter commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+        refreshInterval.foreach { interval =>
+          val initial =
+            if (interval >= 2.seconds)
+              ThreadLocalRandom.current().nextLong(interval.toMillis).millis
```
Sorry, what do you mean? To replace 2 seconds with 1 second? or divide interval by 2?

patriknw replied:
This is what EventsByPersistenceIdStage does:
```
            val initial =
              if (interval >= 2.seconds)
                (interval / 2) + ThreadLocalRandom.current().nextLong(interval.toMillis / 2).millis
              else interval
```

fancywriter commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+      override def onTimer(timerKey: Any): Unit = {
+        timerKey match {
+          case Continue =>
+            query()
```
Good point, I will think how to improve it.

raboof replied:
(this has since been protected against with the `queryInProgress` guard)

patriknw commented on core/src/main/scala/akka/persistence/cassandra/ConfigSessionProvider.scala:
```diff
@@ -39,15 +37,12 @@ import com.datastax.driver.core.policies.ConstantSpeculativeExecutionPolicy
 class ConfigSessionProvider(system: ActorSystem, config: Config) extends SessionProvider {
 
-  def connect()(implicit ec: ExecutionContext): Future[Session] = {
+  def connect()(implicit ec: ExecutionContextExecutor): Future[Session] = {
```
just curious, what is it that requires `ExecutionContextExecutor`?

patriknw commented on core/src/main/scala/akka/persistence/cassandra/package.scala:
```diff
@@ -32,23 +31,12 @@ import akka.annotation.InternalApi
-
   implicit class ListenableFutureConverter[A](val lf: ListenableFuture[A]) extends AnyVal {
-    def asScala(implicit ec: ExecutionContext): Future[A] = {
+    def asScala(implicit ec: Executor): Future[A] = {
```
ah, this is the reason for the ExecutionContextExecutor. I agree about reverting.

An Akka MesageDispatcher is always an Executor.

patriknw commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+        }
+      }
+
+      override def preStart(): Unit = {
```
sounds good to pre-fetch

patriknw commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+        while (buffer.nonEmpty && isAvailable(out)) {
+          val (s, newBuffer) = buffer.dequeue
+          buffer = newBuffer
+          push(out, s)
```
It will only ever be possible to push one element here, then there will be a new call to `onPull`.
So this should be named `tryPushOne` instead of `pushWhileAvailable`

patriknw commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+          for (_ <- 1 to available) {
+            val s = rs.one().getString("persistence_id")
+            if (!knownPersistenceIds.contains(s)) {
+              buffer = buffer.enqueue(s)
```
ok, good to drain the ResultSet immediately since most persistenceIds will already be known

patriknw commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,130 @@
+            }
+          }
+          pushWhileAvailable()
+          if (refreshInterval.isEmpty && buffer.isEmpty && rs.getExecutionInfo.getPagingState == null) {
```
we use `rs.isExhausted()` in other places

ennru commented:
For 1.0 we're planning to use the Cassandra Java driver 4.0.x with its proper async APIs https://github.com/akka/akka-persistence-cassandra/issues/446

chbatey commented:
> For 1.0 we're planning to use the Cassandra Java driver 4.0.x with its proper async APIs #446

I think we should carefully test their non blocking semantics

ennru commented:
As we're preparing for a 1.0 and to stay compatible with Akka 2.6, I'd like to get your suggested changes in. 
I've rolled back the change to avoid the `Executor` cast and did some changes suggested by @chbatey and @patriknw.
The diff is much smaller now, but with all the changes to master we need to decide how to rebase it.

ennru commented:
To get to the essence of the PR, I've reverted the future conversions changes.

chbatey commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,135 @@
+              buffer = buffer.enqueue(s)
+              knownPersistenceIds += s
+            }
+          }
```
shouldn't we trigger fetch more results if it isn't exhaisted?

raboof replied:
This is now done in `onPull`

chbatey commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,135 @@
+                  query()
+
+                case Some(rs) =>
+                  if (rs.isExhausted) {
```
if it is exhausted we fetch more results, doesn't seem right

ennru replied:
Oops, I turned the criterion around...

ennru commented:
It seems like no tests rely on paging the result set.

fancywriter commented:
@ennru I understood why you have reverted implicit executor parameter change (still don't like it, but OK if it is public API). But why have you reverted my change from implicit conversion to explicit (ListenableFuture to scala Future)? Both of them are currently used in project, I wanted to make it consistent, so to you only explicit conversion and drop implicit one.

If you are OK with this change, I will bring it back and rebase with latest master, what do you think?

fancywriter commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,132 @@
+        materializer.executionContext
+
+      private var queryInProgress = false
+      private var knownPersistenceIds: Set[String] = Set.empty
```
@patriknw @chbatey @ennru what do you think about using immutables set and queue here? isn't it better to replace them with mutable versions? It is not public API anyway, it's internal private state. I firstly thought that immutable collections are more idiomatic and safe to use, but now I think it could affect performance. Do you have any performance tests for this query?

raboof replied:
I don't think we have any performance tests for this stage.

The operations we do on this set are 'add' and 'contains'. For those I don't see a reason to assume a mutable set would be faster: it might even be slower since it'd have to protect against concurrent access.

It does seem a bit wasteful to have all persistence id's in memory like this, it could be interesting to experiment with other implementations. That seems outside the scope of this PR though - the previous implementation also used an immutable `Set` so this PR shouldn't introduce a regression here.

patriknw replied:
This query probably has much worse scalability problems, such as the actual Cassandra query. Let's not be too ambitious here, it's probably not used for anything "real".

ennru commented:
@fancywriter I wanted to boil this PR down to just replacing the `QueryActorPublisher` so that it could easily be merged to both 0.x and master.

We first had plans to rush the Akka Persistence Cassandra 1.0 a bit, but then understood that we'd want to use the new Cassandra Java client API which will have a large impact. This work is now postponed until after Akka 2.6 is final. 

This makes it even more important to get this PR into 0.x, as the Actor Publisher will be removed from Akka 2.6.

fancywriter commented:
@ennru yes, that's what I did already yesterday. And rebased with latest master. Now changes are in only four files - deleted AllPersistenceIdsPublisher and QueryActorPublisher, added new AllPersistenceIdsStage and changed CassanraReadJournal to use stage instead of actor publisher.

TimMoore commented:
Related suggestion: #518 — create another CI job that tests with Akka 2.6 builds.

raboof commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,128 @@
+          session.executeAsync(boundStatement).asScala.foreach(queryCallback.invoke)
+        }
+
+      private def tryPushOne(): Unit = {
```
Let's rename this (to something like `flush`?) as it actually tries to push as many elements as it can as long as the buffer is nonempty and the outlet is available.

raboof commented on core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala:
```diff
@@ -0,0 +1,128 @@
+              if (refreshInterval.isEmpty && rs.isExhausted) {
+                complete(out)
+              } else {
+                rs.fetchMoreResults().asScala.foreach(queryCallback.invoke)
```
Should this also be guarded by `queryInProgress` to avoid calling `fetchMoreResults` in parallel?

chbatey replied:
Yep I think so, I'll add that change then merge this!

chbatey commented:
I committed your changes in PR https://github.com/akka/akka-persistence-cassandra/pull/520 (keeping credits). Thanks @fancywriter !
